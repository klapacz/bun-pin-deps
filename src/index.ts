import { parseModule, generateCode } from "magicast";

type Deps = Record<string, string>;

export function pinDeps({ code: rawCode, deps }: { code: string; deps: Deps }) {
  const mod = parseModule(rawCode);

  for (const key in mod.imports) {
    const value = mod.imports[key];
    if (!Object.keys(deps).includes(value.from)) {
      continue;
    }

    value.from = `${value.from}@${deps[value.from as keyof typeof deps]}`;
  }

  const { code } = generateCode(mod);
  return code;
}
