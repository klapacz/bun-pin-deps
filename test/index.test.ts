import { describe, it, expect } from "bun:test";
import { pinDeps } from "../src";

describe("should", () => {
  it("pin external dependency", () => {
    const input = `import { z } from "zod"; const number = z.number().parse(1);`;
    const output = pinDeps({ code: input, deps: { zod: "^3.21.4" } });

    expect(output).toMatchSnapshot();
  });
});
