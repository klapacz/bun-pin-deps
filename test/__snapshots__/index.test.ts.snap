// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should pin external dependency 1`] = `"import {z} from "zod@^3.21.4"; const number = z.number().parse(1);"`;
